### ISSUE TITLE
Storing (arbitrary) data in metadata

### ISSUE THREAD

**Issue by sebastian-goeldi** [KLayout/klayout#1341]
Hi Matthias,

I am creating this ticket to track what we talked about ;) (I hope I am mot misremembering or straight up telling wrong things)

It would be nice if we can store metadata of cells (and maybe instance) in metadata. This would enable us to use Klayout for conecepts such as ports in photonics. My poor understanding is that this is how PCells currently handle this and it could be generalized to enable heneric access to the metadata.

Feel free to edit/delete/recreate however you see fit.

@joamatab @gdspaul

**Comment by klayoutmatthias** [comment]
Thanks :) I have planned to include it in 0.28.8. 

Matthias

**Comment by EugeneZelenko** [comment]
@sebastian-goeldi: Did you try to use user properties? They allow to store arbitrary values.

**Comment by sebastian-goeldi** [comment]
> @sebastian-goeldi: Did you try to use user properties? They allow to store arbitrary values.

Hi Eugene,

Yes I am totally aware. The problem there is that user properties are not as flexible as the MetaInfo. Due to gds limitations, I would need to do a manual mapping of index <-> object. This way allows more flexibility.

**Comment by klayoutmatthias** [comment]
@EugeneZelenko meta info can store arbitrary data also in GDS which is otherwise limited to strings and numeric keys. They also support complex objects and they do not pollute the user property space. Similar to user properties but better.

**Comment by EugeneZelenko** [comment]
@klayoutmatthias: Thank you for clarification!

**Comment by sebastian-goeldi** [comment]
@klayoutmatthias 

Hi Matthias,

I saw that you merged test cases and the enhancement into `master`. Because I am an impatient waiter, I have tried it out with running the python module compiled from `master`. What I notices is that even if the LayoutMetaInfo added to a Cell with `add_meta_info` (even with `persisted = True`, the Cell reports it properly, but when saving the layout, it doesn't save (or load not sure how to find out). Is that not implemented yet, or is there something missing in the code to make that work?

Lmk if I just need to be a bit more patient, or if there is something that I can help with. ^^

Best,
Sebastian
